Configs are Mutated by File-Level Overrides

### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.


### What Happened

When using a `FluffConfig` object as a config when linting a SQL string, any file-level overrides will mutate the config object, causing subsequent lintings to use the same file-level override. 

### Expected Behaviour

```python
def test__api__immutable_config():
    config = FluffConfig.from_path("test/fixtures/api/config_path_test/extra_configs/.sqlfluff")
    assert config.get("dialect") == "ansi"
    sqlfluff.lint(
        "-- sqlfluff:dialect: postgres\n"
        "SELECT * FROM table1\n",
        config=config
    )
    assert config.get("dialect") == "ansi"  # dialect should be unchanged
```

### Observed Behaviour

The test above fails:
`AssertionError: assert 'postgres' == 'ansi'`
The file-level override mutates the config object.

### How to reproduce

This test reproduces the issue:
```python
def test__api__immutable_config():
    config = FluffConfig.from_path("test/fixtures/api/config_path_test/extra_configs/.sqlfluff")
    assert config.get("dialect") == "ansi"
    sqlfluff.lint(
        "-- sqlfluff:dialect: postgres\n"
        "SELECT * FROM table1\n",
        config=config
    )
    assert config.get("dialect") == "ansi"
```

### Dialect

Ansi & Postgres, but this has been observed with every dialect.

### Version

3.2.3

### Configuration

I am using the test config available at `test/fixtures/api/config_path_test/extra_configs/.sqlfluff`.

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

--- a/src/sqlfluff/core/config/fluffconfig.py
+++ b/src/sqlfluff/core/config/fluffconfig.py
@@ -156,7 +156,11 @@ class FluffConfig:
         # the type of a templater in their context, use
         # `get_templater_class()` instead, which avoids instantiating
         # a new templater instance.
-        state["_configs"]["core"].pop("templater_obj", None)
+        # NOTE: It's important that we do this on a copy so that we
+        # don't disturb the original object if it's still in use.
+        state["_configs"] = state["_configs"].copy()
+        state["_configs"]["core"] = state["_configs"]["core"].copy()
+        state["_configs"]["core"]["templater_obj"] = None
         return state
 
     def __setstate__(self, state: Dict[str, Any]) -> None:  # pragma: no cover
@@ -178,6 +182,13 @@ class FluffConfig:
         configs_attribute_copy = deepcopy(self._configs)
         config_copy = copy(self)
         config_copy._configs = configs_attribute_copy
+        # During the initial `.copy()`, we use the same `__reduce__()` method
+        # which is used during pickling. The `templater_obj` doesn't pickle
+        # well so is normally removed, but it's ok for us to just pass across
+        # the original object here as we're in the same process.
+        configs_attribute_copy["core"]["templater_obj"] = self._configs["core"][
+            "templater_obj"
+        ]
         return config_copy
 
     @classmethod

--- a/src/sqlfluff/core/config/fluffconfig.py
+++ b/src/sqlfluff/core/config/fluffconfig.py
@@ -3,6 +3,7 @@
 from __future__ import annotations
 
 import logging
+from copy import copy, deepcopy
 from itertools import chain
 from typing import (
     TYPE_CHECKING,
@@ -167,6 +168,17 @@ class FluffConfig:
         # NOTE: Likewise we don't reinstate the "templater_obj" config value
         # which should also only be used in the main thread rather than child
         # processes.
+
+    def copy(self) -> FluffConfig:
+        """Returns a copy of the FluffConfig.
+
+        This creates a shallow copy of most of the config, but with a deep copy of the
+        `_configs` dictionary.
+        """
+        configs_attribute_copy = deepcopy(self._configs)
+        config_copy = copy(self)
+        config_copy._configs = configs_attribute_copy
+        return config_copy
 
     @classmethod
     def from_root(

--- a/src/sqlfluff/core/linter/linter.py
+++ b/src/sqlfluff/core/linter/linter.py
@@ -896,7 +896,7 @@ class Linter:
             self.formatter.dispatch_template_header(fname, self.config, config)
 
         # Just use the local config from here:
-        config = config or self.config
+        config = (config or self.config).copy()
 
         # Scan the raw file for config commands.
         config.process_raw_file_for_config(in_str, fname)

--- a/test/core/config/fluffconfig_test.py
+++ b/test/core/config/fluffconfig_test.py
@@ -5,6 +5,7 @@ import os
 
 import pytest
 
+import sqlfluff
 from sqlfluff.core import FluffConfig, Linter
 from sqlfluff.core.errors import SQLFluffUserError
 from sqlfluff.core.templaters import (
@@ -374,3 +375,15 @@ def test__process_raw_file_for_config(raw_sql):
     # internal list attributes should have overridden exploded list values
     assert cfg.get("rule_allowlist") == ["LT05", "LT06"]
     assert cfg.get("rule_denylist") == ["LT01", "LT02"]
+
+
+def test__api__immutable_config():
+    """Tests that a config is not mutated when parsing."""
+    config = FluffConfig.from_path(
+        "test/fixtures/api/config_path_test/extra_configs/.sqlfluff"
+    )
+    assert config.get("dialect") == "ansi"
+    sqlfluff.parse(
+        "-- sqlfluff:dialect: postgres\nSELECT * FROM table1\n", config=config
+    )
+    assert config.get("dialect") == "ansi"

